File: 백준/Silver/9012. 괄호/괄호.py
import sys

input = sys.stdin.readline

n = int(input())

_OPEN = "("
_CLOSE = ")"
_YES = "YES"
_NO = "NO"

for _ in range(n) :
    parenthesis = list(input().strip())
    stack = []
    result = _YES
    for el in parenthesis:
        if el == _OPEN:
            stack.append(el)

        else:
            if len(stack) == 0:
                result = _NO
                break
            else:
                stack.pop()

    if len(stack) != 0:
        result = _NO
    else:
        if result != _NO:
            result = _YES
    print(result)

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int T = Integer.parseInt(br.readLine());

        for (int i = 0; i < T; i++) {
            Stack<Character> stack = new Stack<>();
            String inputStr = br.readLine();
            boolean flag = false;
            for (char s : inputStr.toCharArray()) {
                if (s == '(') {
                    stack.push(s);
                    continue;
                }

                if (stack.isEmpty()) {
                    System.out.println("NO");
                    flag = true;
                    break;
                }

                stack.pop();
            }

            if (!flag && stack.isEmpty()) {
                System.out.println("YES");
                continue;
            }

            if (!flag) {
                System.out.println("NO");
            }

        }
    }
}